packages/web/queries/complex/pools.ts

```diff
@@ -0,0 +1,325 @@
+import { queryBalances } from "~/queries/cosmos";
+import {
+  FilteredPoolsResponse,
+  PoolToken,
+  queryFilteredPools,
+} from "~/queries/indexer";
+import { queryNumPools, queryPools } from "~/queries/osmosis";
+
+export type PoolRaw =
+  | CosmwasmPoolRaw
+  | StablePoolRaw
+  | ConcentratedLiquidityPoolRaw
+  | WeightedPoolRaw;
+
+export async function queryPaginatedPools(
+  page?: number,
+  limit?: number,
+  poolIdParam?: string
+): Promise<{ status: number; pools: PoolRaw[] }> {
+  // Fetch the pools data from your database or other source
+  // This is just a placeholder, replace it with your actual data fetching logic
+  const allPools: PoolRaw[] = await fetchAndProcessAllPools();
+
+  // Handle the case where specific pool ID is requested
+  if (poolIdParam) {
+    const pool = allPools.find(
+      (pool) => ("pool_id" in pool ? pool.pool_id : pool.id) === poolIdParam
+    );
+    if (!pool) {
+      throw { status: 404, pools: [] };
+    }
+    return { status: 200, pools: [pool] };
+  }
+
+  // Pagination
+  if (page && limit) {
+    const startIndex = (page - 1) * limit;
+
+    // Slice the data based on the page and limit
+    const pools = allPools.slice(startIndex, startIndex + limit);
+
+    // Return the paginated data
+    return { status: 200, pools };
+  }
+
+  // Return the paginated data
+  return { status: 200, pools: allPools };
+}
+
+/** Cache on this current edge function instance. */
+const allPoolsLruCache = new LRUCache<string, CacheEntry>({
+  max: 2,
+});
+
+async function fetchAndProcessAllPools(): Promise<PoolRaw[]> {
+  return cachified({
+    key: "all-pools",
+    cache: allPoolsLruCache,
+    async getFreshValue() {
+      const numPools = await queryNumPools();
+
+      // Fetch all pools from imperator, except cosmwasm pools for now
+      // TODO remove when indexer returns cosmwasm pools
+      try {
+        const filteredPoolsResponse = await queryFilteredPools(
+          {
+            min_liquidity: 0,
+            order_by: "desc",
+            order_key: "liquidity",
+          },
+          { offset: 0, limit: Number(numPools.num_pools) }
+        );
+        const queryPoolRawPromises = filteredPoolsResponse.pools.map(
+          queryPoolRawFromFilteredPool
+        );
+        const queryPoolRawResults = await Promise.all(queryPoolRawPromises);
+
+        const cosmwasmPools = await getCosmwasmPools();
+
+        // prepend cosmwasm pools
+        return (cosmwasmPools as PoolRaw[]).concat(
+          queryPoolRawResults.filter(
+            (
+              poolRaw
+            ): poolRaw is
+              | StablePoolRaw
+              | ConcentratedLiquidityPoolRaw
+              | WeightedPoolRaw => !!poolRaw
+          )
+        );
+      } catch (e) {
+        console.error(e);
+
+        // fall back to pools query on node
+        return await getPoolsFromNode();
+      }
+    },
+    ttl: 30 * 1000, // 30 seconds
+  });
+}
+
+export async function queryPoolRawFromFilteredPool(
```

Comment:
We can drop the `async` keyword from the function because there's no `await` used inside. Even with `Promise.all`, the operations will be processed in sequence, not concurrently, due to the absence of `await`. If our goal is true parallel processing, we'd have to leverage the [worker_threads](https://nodejs.org/api/worker_threads.html) module in Node.js. Node.js is inherently single-threaded, processing one task at a time.